pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/BaseContext.java

```diff
@@ -217,4 +217,10 @@ default ClientBuilder getPulsarClientBuilder() {
         throw new UnsupportedOperationException("not implemented");
     }
 
+    /**
+     * Terminate the function instance with a fatal exception.
+     *
+     * @param t the fatal exception to be raised
+     */
+    void fatal(Throwable t);
```

Comment:
When the function/connector calls this method, we don't break user threads to continue executing. Is this reasonable?

For examples: 

```java

    public static class TestSinkConnector implements Sink<String> {
        SinkContext context;

        @Override
        public void open(Map config, SinkContext sinkContext) throws Exception {
            this.context = sinkContext;
        }

        @Override
        public void write(Record<String> record) throws Exception {
            new Thread(() -> {
                if (FailComponentType.FAIL_SINK.toString().equals(record.getValue())) {
                    context.fatal(new Exception(FailComponentType.FAIL_SINK.toString()));
                }
                // continue to do other things.
            }).start();
        }

        @Override
        public void close() throws Exception {

        }
    }
```

Reply:
Yes, it's designed that way. The fatal method does not need to crash the current thread.